File: java/src/main/java/com/newnancity/newnanmanager/services/PlayerServerService.java
package com.newnancity.newnanmanager.services;

import com.newnancity.newnanmanager.SessionContext;
import com.newnancity.newnanmanager.exceptions.NewNanManagerException;
import com.newnancity.newnanmanager.generated.api.PlayerServerServiceApi;
import com.newnancity.newnanmanager.generated.model.PlayerServersResponse;
import com.newnancity.newnanmanager.generated.model.ServerPlayersResponse;
import com.newnancity.newnanmanager.generated.model.SetPlayersOfflineRequest;

/** Player/server relationship and online snapshot service. */
public final class PlayerServerService extends AbstractService {
    private final PlayerServerServiceApi api;

    public PlayerServerService(PlayerServerServiceApi api) {
        this.api = api;
    }

    public PlayerServersResponse getPlayerServers(int playerId) throws NewNanManagerException {
        return getPlayerServers(playerId, null);
    }

    public PlayerServersResponse getPlayerServers(int playerId, Boolean onlineOnly)
            throws NewNanManagerException {
        return execute(() -> api.getPlayerServers(playerId).onlineOnly(onlineOnly).execute());
    }

    public ServerPlayersResponse getServerPlayers() throws NewNanManagerException {
        return getServerPlayers(null, null, null, null, null);
    }

    public ServerPlayersResponse getServerPlayers(
            Integer page,
            Integer pageSize,
            String search,
            Integer serverId,
            Boolean onlineOnly) throws NewNanManagerException {
        return execute(() -> api.getServerPlayers()
                .page(page)
                .pageSize(pageSize)
                .search(search)
                .serverId(serverId)
                .onlineOnly(onlineOnly)
                .execute());
    }

    public void setPlayersOffline(
            String sessionId,
            long sessionEpoch,
            SetPlayersOfflineRequest request) throws NewNanManagerException {
        SessionContext session = new SessionContext(sessionId, sessionEpoch);
        executeVoid(() -> api.setPlayersOffline(session.getId(), session.getEpoch(), request).execute());
    }

    public void setPlayersOffline(
            SetPlayersOfflineRequest request,
            SessionContext session) throws NewNanManagerException {
        setPlayersOffline(session.getId(), session.getEpoch(), request);
    }

    public PlayerServerServiceApi raw() {
        return api;
    }
}

File: kotlin/src/main/kotlin/com/nanmanager/bukkit/services/PlayerServerService.kt
package com.nanmanager.bukkit.services

import com.nanmanager.bukkit.http.HttpClient
import com.nanmanager.bukkit.SessionContext
import com.nanmanager.bukkit.models.*

/** 玩家服务器关系管理服务；传输和模型解析由 OpenAPI 生成客户端负责。 */
class PlayerServerService(httpClient: HttpClient) : BaseService(httpClient) {
    private val api get() = httpClient.generatedApis.playerServers

    fun getPlayerServers(request: GetPlayerServersRequest): PlayerServersResponse {
        return generated<PlayerServersResponse, com.newnanmanager.generated.models.PlayerServersResponse> {
            api.getPlayerServersWithHttpInfo(request.playerId, request.onlineOnly)
        }
    }

    fun getServerPlayers(request: GetServerPlayersRequest): ServerPlayersResponse {
        return generated<ServerPlayersResponse, com.newnanmanager.generated.models.ServerPlayersResponse> {
            api.getServerPlayersWithHttpInfo(
                page = request.page,
                pageSize = request.pageSize,
                search = request.search,
                serverId = request.serverId,
                onlineOnly = request.onlineOnly
            )
        }
    }

    fun setPlayersOffline(request: SetPlayersOfflineRequest) {
        post<Unit>("/api/v1/servers/players/offline", request)
    }

    fun setPlayersOffline(request: SetPlayersOfflineRequest, session: SessionContext) {
        val body = toGenerated<com.newnanmanager.generated.models.SetPlayersOfflineRequest>(request)
        generatedVoid {
            api.setPlayersOfflineWithHttpInfo(session.id, session.epoch, body)
        }
    }
}
